File: sdks/typescript/src/financial-attestation.ts
import type { CommitFinancialActionRequest } from './generated/CommitFinancialActionRequest';

const PREFIX = 'tlg-financial-execution-attestation.v1';
const encoder = new TextEncoder();

export type UnsignedFinancialExecutionAttestation = Omit<
  CommitFinancialActionRequest,
  'signature'
>;

export function financialExecutionAttestationMessage(
  actionId: string,
  request: UnsignedFinancialExecutionAttestation,
): Uint8Array {
  const fields = [
    request.connector_id,
    actionId,
    request.grant_id,
    request.action_hash,
    request.provider,
    request.provider_reference,
    request.provider_status,
    request.executed_at,
    request.idempotency_key,
    request.provider_proof_sha256,
  ];
  const lines = [PREFIX, ...fields.map((field) => `${encoder.encode(field).byteLength}:${field}`)];
  return encoder.encode(lines.join('\n'));
}

export async function financialProviderProofSha256(proof: string): Promise<string> {
  const digest = await crypto.subtle.digest('SHA-256', arrayBuffer(encoder.encode(proof)));
  return `sha256:${hex(new Uint8Array(digest))}`;
}

export async function signFinancialExecutionAttestation(
  plaintextSecret: string,
  actionId: string,
  request: UnsignedFinancialExecutionAttestation,
): Promise<string> {
  const key = await crypto.subtle.importKey(
    'raw',
    arrayBuffer(decodeBase64Url(plaintextSecret)),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign'],
  );
  const signature = await crypto.subtle.sign(
    'HMAC',
    key,
    arrayBuffer(financialExecutionAttestationMessage(actionId, request)),
  );
  return `v1=${encodeBase64Url(new Uint8Array(signature))}`;
}

function decodeBase64Url(value: string): Uint8Array {
  const normalized = value.replace(/-/g, '+').replace(/_/g, '/');
  const padded = normalized.padEnd(Math.ceil(normalized.length / 4) * 4, '=');
  const binary = atob(padded);
  return Uint8Array.from(binary, (character) => character.charCodeAt(0));
}

function encodeBase64Url(value: Uint8Array): string {
  let binary = '';
  for (const byte of value) binary += String.fromCharCode(byte);
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '');
}

function hex(value: Uint8Array): string {
  return Array.from(value, (byte) => byte.toString(16).padStart(2, '0')).join('');
}

function arrayBuffer(value: Uint8Array): ArrayBuffer {
  const copy = new Uint8Array(value.byteLength);
  copy.set(value);
  return copy.buffer;
}

File: sdks/python/src/trustloopguard/financial_attestation.py
"""Canonical signing helpers for external financial execution attestations."""

from __future__ import annotations

import base64
import hashlib
import hmac
from typing import Mapping

PREFIX = "tlg-financial-execution-attestation.v1"


def financial_execution_attestation_message(
    action_id: str, request: Mapping[str, str]
) -> bytes:
    fields = [
        request["connector_id"],
        action_id,
        request["grant_id"],
        request["action_hash"],
        request["provider"],
        request["provider_reference"],
        request["provider_status"],
        request["executed_at"],
        request["idempotency_key"],
        request["provider_proof_sha256"],
    ]
    lines = [PREFIX, *(f"{len(value.encode('utf-8'))}:{value}" for value in fields)]
    return "\n".join(lines).encode("utf-8")


def financial_provider_proof_sha256(proof: str) -> str:
    return f"sha256:{hashlib.sha256(proof.encode('utf-8')).hexdigest()}"


def sign_financial_execution_attestation(
    plaintext_secret: str, action_id: str, request: Mapping[str, str]
) -> str:
    secret = _decode_base64url(plaintext_secret)
    signature = hmac.new(
        secret,
        financial_execution_attestation_message(action_id, request),
        hashlib.sha256,
    ).digest()
    return f"v1={base64.urlsafe_b64encode(signature).decode('ascii').rstrip('=')}"


def _decode_base64url(value: str) -> bytes:
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))
